Research/Research/RSDColorMatrix.swift

```diff
@@ -0,0 +1,231 @@
+    }
+    
+    public func colorMapping(for gray: RSDGrayScale.Shade, version: Int? = nil) -> RSDColorMapping {
+        return self.grayScale(for: version).mapping(for: gray)
+    }
+    
+    public func findColorTile(for color: RSDColor) -> RSDColorTile? {
+        for colorTile in RSDGrayScale().colorTiles {
+            if colorTile.color == color {
+                return colorTile
+            }
+        }
+        for library in self.registeredLibraries.reversed() {
+            for swatch in library.swatches {
+                for colorTile in swatch.colorTiles {
+                    if colorTile.color == color {
+                        return colorTile
+                    }
+                }
+            }
+            if let grayScale = library.grayScale {
+                for colorTile in grayScale.colorTiles {
+                    if colorTile.color == color {
+                        return colorTile
+                    }
+                }
+            }
+        }
+        return nil
+    }
+}
+
+
+/// A color library is a versioned collection of color swatches.
+public struct RSDColorLibrary : Codable, Equatable, Hashable, Comparable {
+
+    /// The version for this library.
+    public var version: Int
+    
+    /// A gray scale for the color library.
+    public var grayScale: RSDGrayScale?
+    
+    /// A list of all the color swatches included in this color library.
+    public var swatches: Set<RSDColorSwatch>
+    
+    /// Get the color swatch associated with the given name.
+    public func colorSwatch(for name: String) -> RSDColorSwatch? {
+        return self.swatches.first(where: { $0.name == name })
+    }
+    
+    /// Get the color swatch associated with the given name.
+    public func colorSwatch(for name: RSDReservedColorName) -> RSDColorSwatch? {
+        return self.colorSwatch(for: name.rawValue)
+    }
+    
+    public static func < (lhs: RSDColorLibrary, rhs: RSDColorLibrary) -> Bool {
+        return lhs.version < rhs.version
+    }
+}
+
+
+/// A listing of the color names that are already assigned by the base framework using "ColorMatrix.json" as
+/// the source file.
+public enum RSDReservedColorName : RawRepresentable, Equatable, Hashable, Codable {
+    
+    case pallette(Pallette)
+    public enum Pallette : String, CaseIterable {
+        case royal
+        case butterscotch
+        case powder
+        case turquoise
+        case coral
+        case apricot
+        case blueberry
+        case apple
+        case rose
+        case lavender
+        case slate
+        case fern
+        case cloud
+        case stone
+    }
+    
+    case special(Special)
+    public enum Special : String, CaseIterable {
+        case errorRed
+        case successGreen
+    }
+    
+    /// Maximum version in which a given reserved color name is expected to be included. If `nil` this color
+    /// name should still be included in the color matrix in the most recent library.
```

Comment:
The RSDColorMatrix is set up as a final class which uses an embedded JSON file to define the colors.

The idea behind versioning is that colors might be added/removed but that this will never use custom names. This method was added, therefore, to allow for a scenario where someone is decoding a color and wants to first check the built in colors before looking elsewhere for a custom definition.

That said, it  doesn't really add anything, and I can remove the method. :)